File: app/src/main/java/br/com/popularmovies/moviedetail/trailers/ui/MovieTrailerFragment.java
package br.com.popularmovies.moviedetail.trailers.ui;

import android.app.AlertDialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.Group;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import br.com.popularmovies.R;
import br.com.popularmovies.base.interfaces.IConection;
import br.com.popularmovies.data.model.ErrorResponse;
import br.com.popularmovies.data.model.Resource;
import br.com.popularmovies.moviedetail.trailers.adapters.TrailerAdapter;
import br.com.popularmovies.moviedetail.trailers.adapters.TrailerClickListener;
import br.com.popularmovies.moviedetail.trailers.viewmodel.MovieTrailerViewModel;
import br.com.popularmovies.moviedetail.trailers.viewmodel.factories.MovieTrailerFactory;
import br.com.popularmovies.services.movieService.response.MovieTrailers;
import br.com.popularmovies.services.movieService.source.MovieRepository;
import br.com.popularmovies.services.movieService.source.local.MovieLocalDataSource;
import br.com.popularmovies.services.movieService.source.remote.MovieRemoteDataSource;

import static br.com.popularmovies.data.Constants.NETWORK_ERROR_CODE;
import static br.com.popularmovies.movies.Constants.GENERIC_MSG_ERROR_TITLE;
import static br.com.popularmovies.movies.Constants.MOVIE_ID;

public class MovieTrailerFragment extends Fragment implements IConection, TrailerClickListener {

    private MovieTrailerViewModel mViewModel;
    private RecyclerView mTrailersRecyclerView;
    private Observer<Resource<MovieTrailers>> trailersObserver;
    private Group mNoConnectionGroup;
    private Button mTryAgainButton;
    private TextView mNoConnectionText;
    private ProgressBar mProgressBar;
    private TextView mNoTrailers;

    public static MovieTrailerFragment newInstance(int movieId) {
        MovieTrailerFragment movieTrailerFragment = new MovieTrailerFragment();
        Bundle args = new Bundle();
        args.putInt(MOVIE_ID, movieId);
        movieTrailerFragment.setArguments(args);
        return movieTrailerFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setupObservers();
    }

    private void setupObservers() {
        trailersObserver = new Observer<Resource<MovieTrailers>>() {
            @Override
            public void onChanged(@Nullable Resource<MovieTrailers> movieReviewsResource) {
                if (movieReviewsResource != null)
                    switch (movieReviewsResource.status) {
                        case LOADING:
                            showLoading();
                            mTrailersRecyclerView.setVisibility(View.GONE);
                            break;
                        case SUCCESS:
                            hideLoading();
                            mTrailersRecyclerView.setVisibility(View.VISIBLE);
                            if (movieReviewsResource.data != null) {
                                if (movieReviewsResource.data.getTrailers().isEmpty()) {
                                    showNoTrailers();
                                } else {
                                    TrailerAdapter mTrailerAdapter = new TrailerAdapter(movieReviewsResource.data.getTrailers(), MovieTrailerFragment.this);
                                    mTrailersRecyclerView.setAdapter(mTrailerAdapter);
                                    showResult();

                                }
                            }
                            break;
                        case ERROR:
                            hideLoading();
                            ErrorResponse error = movieReviewsResource.error;
                            if (error != null) {
                                if (error.getStatusCode() == NETWORK_ERROR_CODE) {
                                    showNoConnection(error.getStatusMessage());
                                } else {
                                    showGenericError(error.getStatusMessage());
                                }
                            }
                            break;
                    }
            }
        };
    }

    private void showNoTrailers() {
        mNoTrailers.setVisibility(View.VISIBLE);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.movie_trailer_fragment, container, false);
        Bundle args = getArguments();
        if (args != null) {
            int movieId = args.getInt(MOVIE_ID, -1);
            MovieRepository mMovieRepository = MovieRepository.getInstance(MovieLocalDataSource.getInstance(requireActivity().getApplicationContext())
                    , MovieRemoteDataSource.getInstance());
            mViewModel = ViewModelProviders.of(this,
                    new MovieTrailerFactory(mMovieRepository, movieId)).get(MovieTrailerViewModel.class);
            setupFields(view);
            setupTrailersList(view);
            mViewModel.getTrailers().observe(getViewLifecycleOwner(), trailersObserver);
            mTryAgainButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    tryAgain();
                }
            });
        }

        return view;
    }

    private void setupFields(View view) {
        mNoConnectionGroup = view.findViewById(R.id.group_no_connection);
        mNoConnectionText = view.findViewById(R.id.tv_no_conection);
        mTryAgainButton = view.findViewById(R.id.bt_try_again);
        mProgressBar = view.findViewById(R.id.pb_base);
        mNoTrailers = view.findViewById(R.id.tv_no_trailers);
    }

    private void setupTrailersList(View view) {
        mTrailersRecyclerView = view.findViewById(R.id.rv_trailers);
        mTrailersRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    }

    @Override
    public void showLoading() {
        mProgressBar.setVisibility(View.VISIBLE);
        mNoConnectionGroup.setVisibility(View.GONE);
    }

    @Override
    public void hideLoading() {
        mProgressBar.setVisibility(View.GONE);
    }

    @Override
    public void showResult() {
        mNoConnectionGroup.setVisibility(View.GONE);
        mNoTrailers.setVisibility(View.GONE);
    }

    @Override
    public void showNoConnection(String message) {
        mTrailersRecyclerView.setVisibility(View.GONE);
        mNoConnectionText.setText(message);
        mNoConnectionGroup.setVisibility(View.VISIBLE);
    }

    @Override
    public void showGenericError(String message) {
        final AlertDialog sortDialog = new AlertDialog.Builder(getContext())
                .setTitle(GENERIC_MSG_ERROR_TITLE)
                .setMessage(message)
                .setPositiveButton(R.string.dialog_ok, null)
                .create();

        sortDialog.show();
    }

    @Override
    public void tryAgain() {
        mViewModel.tryAgain();
    }

    @Override
    public void onPlay(String videoUrl) {
        Toast.makeText(requireContext(), "onPlay", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onShare(String videoSite) {
        Toast.makeText(requireContext(), "onShare", Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/br/com/popularmovies/moviedetail/trailers/ui/MovieTrailerFragment.kt
package br.com.popularmovies.moviedetail.trailers.ui

import android.app.AlertDialog
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.app.ShareCompat
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.fragment.navArgs
import br.com.popularmovies.R
import br.com.popularmovies.appComponent
import br.com.popularmovies.base.interfaces.IConection
import br.com.popularmovies.common.configs.ErrorCodes.NETWORK_ERROR_CODE
import br.com.popularmovies.common.configs.ErrorMessages.GENERIC_MSG_ERROR_TITLE
import br.com.popularmovies.databinding.MovieTrailerFragmentBinding
import br.com.popularmovies.moviedetail.trailers.Constants.YOUTUBE_URL
import br.com.popularmovies.moviedetail.trailers.adapters.TrailerAdapter
import br.com.popularmovies.moviedetail.trailers.adapters.TrailerClickListener
import br.com.popularmovies.moviedetail.trailers.viewmodel.MovieTrailerViewModel
import kotlinx.android.synthetic.main.movie_trailer_fragment.*

class MovieTrailerFragment : Fragment(), IConection, TrailerClickListener {

    private val args by navArgs<MovieTrailerFragmentArgs>()

    private val mViewModel: MovieTrailerViewModel by lazy {
        appComponent.movieTrailerViewModelFactory.create(args.movieId)
    }

    private lateinit var binding: MovieTrailerFragmentBinding

    override fun onAttach(context: Context) {
        super.onAttach(context)

        val movieDetailComponent = appComponent.movieDetailComponent().create()
        movieDetailComponent.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupObservers()
    }

    private fun setupObservers() {
        setupMovieTrailersObserver()
        setupErrorObserver()
        setupLoadingObserver()
    }

    private fun setupLoadingObserver() {
        mViewModel.loading.observe(this, Observer { status ->
            if (status == true) {
                binding.rvTrailers.visibility = View.GONE
            } else {
                hideLoading()
            }
        })
    }

    private fun setupErrorObserver() {
        mViewModel.error.observe(this, Observer { error ->
            mViewModel.showLoading(false)
            if (error != null) {
                if (error.code == NETWORK_ERROR_CODE) {
                    showNoConnection(error.message)
                } else {
                    showGenericError(error.message)
                }
            }
        })
    }

    private fun setupMovieTrailersObserver() {
        mViewModel.trailers.observe(this, Observer { movieTrailers ->
            mViewModel.showLoading(false)
            if (movieTrailers.isEmpty()) {
                showNoTrailers()
            } else {
                binding.rvTrailers.adapter = TrailerAdapter(
                        movieTrailers,
                        this@MovieTrailerFragment
                )
                showResult()
            }
        })
    }

    private fun showNoTrailers() {
        binding.tvNoTrailers.visibility = View.VISIBLE
    }

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        super.onCreateView(inflater, container, savedInstanceState)
        binding = DataBindingUtil.inflate(inflater, R.layout.movie_trailer_fragment, container, false)
        binding.lifecycleOwner = this
        binding.viewModel = mViewModel
        binding.iBaseLayout.btTryAgain.setOnClickListener { tryAgain() }
        return binding.root
    }

    override fun showLoading() {
        binding.iBaseLayout.pbBase.visibility = View.VISIBLE
        binding.iBaseLayout.groupNoConnection.visibility = View.GONE
    }

    override fun hideLoading() {
        binding.iBaseLayout.pbBase.visibility = View.GONE
    }

    override fun showResult() {
        binding.iBaseLayout.groupNoConnection.visibility = View.GONE
        binding.tvNoTrailers.visibility = View.GONE
        binding.rvTrailers.visibility = View.VISIBLE
    }

    override fun showNoConnection(message: String) {
        binding.rvTrailers.visibility = View.GONE
        binding.iBaseLayout.tvNoConection.text = message
        binding.iBaseLayout.groupNoConnection.visibility = View.VISIBLE
    }

    override fun showGenericError(message: String) {
        val sortDialog = AlertDialog.Builder(context)
                .setTitle(GENERIC_MSG_ERROR_TITLE)
                .setMessage(message)
                .setPositiveButton(R.string.dialog_ok, null)
                .create()

        sortDialog.show()
    }

    override fun tryAgain() {
        mViewModel.tryAgain()
    }

    override fun onPlay(videoKey: String) {
        val appIntent = Intent(Intent.ACTION_VIEW, Uri.parse("vnd.youtube:$videoKey"))
        val webIntent = Intent(
                Intent.ACTION_VIEW,
                Uri.parse(YOUTUBE_URL + videoKey)
        )
        try {
            requireContext().startActivity(appIntent)
        } catch (ex: ActivityNotFoundException) {
            requireContext().startActivity(webIntent)
        }
    }

    override fun onShare(videoUrl: String) {
        val mimeType = "text/plain"

        val title = "Sharing this trailer on"

        ShareCompat.IntentBuilder
                .from(requireActivity())
                .setType(mimeType)
                .setChooserTitle(title)
                .setText(videoUrl)
                .startChooser()
    }
}
